core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java

```diff
@@ -966,6 +967,13 @@ public static Expression convert(Expression operand, Type fromType,
     if (fromType.equals(toType)) {
       return operand;
     }
+    if (toType instanceof Types.RecordType) {
+      // We can't extract Class from RecordType since mapping Java Class might not generated yet.
+      return operand;
+    }
+    if (Types.isAssignableFrom(toType, fromType)) {
+      return operand;
+    }
     // E.g. from "Short" to "int".
     // Generate "x.intValue()".
```

Comment:
Can we define a new method named `needsCast(Type tpe1, Type tpe2)` in `Types` ? Then `Types.castIfNecessary` and `RexToLixTranslator.convert` can both reuse this method. The method should include the content:
```java
    if (fromType.equals(toType)) {
      return operand;
    }
    if (toType instanceof Types.RecordType) {
      // We can't extract Class from RecordType since mapping Java Class might not generated yet.
      return operand;
    }
    if (Types.isAssignableFrom(toType, fromType)) {
      return operand;
    }
```

Reply:
Good idea, I will put them into `Types` for reuse.